crates/ra_hir/src/ty.rs

```diff
@@ -357,6 +366,14 @@ impl Ty {
         });
         self
     }
+
+    fn builtin_deref(&self) -> Option<Ty> {
+        match self {
+            Ty::Ref(t, _) => Some(Ty::clone(t)),
+            Ty::RawPtr(t, _) => Some(Ty::clone(t)),
```

Comment:
This is `unsafe`, right? Is this the place to check that sort of thing?

Reply:
Hm, I'm not sure actually. We'll generally need some way of recording diagnostics during type checking, and this might fit in with that.

Reply:
Yeah, "how we do diagnostics" is a big open question. There's a tension between:

* a universal diagnostics struct is convenient for consumers
* specific diagnostic struct for each error is convenient for producer
* ide would like to suggest fixes to diagnostics, so it needs to be able to get specific infos somehow.